x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authc/AuthenticationTests.java

```diff
@@ -7,26 +7,150 @@
+
 public class AuthenticationTests extends ESTestCase {
 
     public void testWillGetLookedUpByWhenItExists() {
-        final Authentication.RealmRef authenticatedBy = new Authentication.RealmRef("auth_by", "auth_by_type", "node");
-        final Authentication.RealmRef lookedUpBy = new Authentication.RealmRef("lookup_by", "lookup_by_type", "node");
+        final RealmRef authenticatedBy = new RealmRef("auth_by", "auth_by_type", "node");
+        final RealmRef lookedUpBy = new RealmRef("lookup_by", "lookup_by_type", "node");
         final Authentication authentication = new Authentication(
             new User("user"), authenticatedBy, lookedUpBy);
 
         assertEquals(lookedUpBy, authentication.getSourceRealm());
     }
 
     public void testWillGetAuthenticateByWhenLookupIsNull() {
-        final Authentication.RealmRef authenticatedBy = new Authentication.RealmRef("auth_by", "auth_by_type", "node");
+        final RealmRef authenticatedBy = new RealmRef("auth_by", "auth_by_type", "node");
         final Authentication authentication = new Authentication(
             new User("user"), authenticatedBy, null);
 
         assertEquals(authenticatedBy, authentication.getSourceRealm());
     }
 
+    public void testSameUserAsForRealms() {
+        final User user1 = randomUserOtherThan(new User[0]);
+        final User user2 = randomUserOtherThan(user1);
+
+        final RealmRef fileRealm1 = new RealmRef(
+            randomAlphaOfLengthBetween(3, 8), FileRealmSettings.TYPE, randomAlphaOfLengthBetween(3, 8));
+        final RealmRef fileRealm2 = new RealmRef(
+            randomValueOtherThan(fileRealm1.getName(), () -> randomAlphaOfLengthBetween(3, 8)),
+            FileRealmSettings.TYPE, randomAlphaOfLengthBetween(3, 8));
+
+        final RealmRef nativeRealm1 = new RealmRef(
+            randomAlphaOfLengthBetween(3, 8), NativeRealmSettings.TYPE, randomAlphaOfLengthBetween(3, 8));
+        final RealmRef nativeRealm2 = new RealmRef(
+            randomValueOtherThan(nativeRealm1.getName(), () -> randomAlphaOfLengthBetween(3, 8)),
+            NativeRealmSettings.TYPE, randomAlphaOfLengthBetween(3, 8));
+
+        final RealmRef realm1 = randomNonFileNativeRealmOtherThan(new RealmRef[0]);
+        final RealmRef realm1Alike = new RealmRef(
+            randomValueOtherThan(realm1.getName(), () -> randomAlphaOfLengthBetween(3, 8)),
+            realm1.getType(), randomAlphaOfLengthBetween(3, 8));
+        final RealmRef realm2 = randomNonFileNativeRealmOtherThan(realm1, realm1Alike);
+        final RealmRef realm3 = randomNonFileNativeRealmOtherThan(realm1, realm1Alike, realm2);
+        final RealmRef realm4 = randomNonFileNativeRealmOtherThan(realm1, realm1Alike, realm2, realm3);
+        final RealmRef authRealm = randomFrom(realm2, realm3, realm4);
+        final RealmRef lookupRealm = randomFrom(realm2, realm3, realm4, null);
```

Comment:
I think the test would benefit if you randomize the realms that take part in the `sameUser` check to sometimes be authn realms sometimes lookup realms.
I haven't looked, but there should be a test helper method for this, and if there isn't maybe you can add it.

Reply:
The tests are a bit of mess. Thanks for the suggestion. I basically re-wrote it. Hopefully it is now easier to read.